Assets/Scripts/GameAI/GameObjects/AIGameObject.cs

```diff
@@ -114,21 +134,55 @@ public abstract class AIGameObject : MonoBehaviour
         public RigidbodyConstraints DefaultConstraints { get; private set; }
 
         [HideInInspector]
-        public bool targetInLineOfSight = false;
+        public bool isAggroed = false;
 
         protected Vector3 moveDirection = Vector3.zero;
         protected Vector3 rotationDirection = Vector3.zero;
         private Vector3 newVelocity = Vector3.zero;
         private Vector3 velocityChange = Vector3.zero;
         private float prevYVel = 0;
 
+        //Additional movement forces that make an agent attempt to keep away from other agents and obstacles.
+        protected Vector3 collisionAvoidanceForce = Vector3.zero;
+        protected Vector3 obstacleAvoidanceForce = Vector3.zero;
+
+        //Multiplier for the collision avoidance force for this specific enemy
+        public float individualCollisionAvoidanceModifier = 1.0f;
+
+        //Multiplier for the collision avoidance force for this specific enemy
+        public float individualCollisionAvoidanceMaxDistance;
+
+        //Adjusted avoidance forces that are decreased as the angle between them and the movement direction decreases.
+        protected Vector3 adjustedCollisionAvoidanceForce = Vector3.zero;
+        protected Vector3 adjustedObstacleAvoidanceForce = Vector3.zero;
+
+        //Private variables used in the AdjustAvoidanceForceBasedOnMovementVelocity to store intermediate results.
+        private Vector3 adjustedForce;
+        private Vector3 difference;
+
+        public bool permissionToAttack = false;
+        public bool isAttacking = false;
+
+        public bool debugFlocking = false;
+        public bool debugEngage = false;
+
         public virtual void Init()
         {
             if (rb == null)
             {
                 rb = GetComponent<Rigidbody>();
             }
 
+            if (agentCollider == null)
+            {
+                agentCollider = GetComponent<Collider>();
+            }
+
+            if (strafeHitBoxes != null)
+            {
+                strafeHitBoxes.Init();
```

Comment:
This is another case where I'm tempted to say just let the class initialize itself on startup. I like the general idea of everything flowing from higher level functions, but since these already need to be MonoBehaviors....

Reply:
Leaving this in because I want direct control over script execution order.